src/direct-binary-upload-process.js

```diff
@@ -12,238 +12,159 @@ governing permissions and limitations under the License.
+ * * fileprogress: Sent periodically to report on how much of the file has
+ *   uploaded. The event data will include these additional elements:
+ *   * {number} transferred: The total number of bytes that have transferred for
+ *     the file so far.
+ * * fileend: Sent after the last part of a file has uploaded successfully. Will only
+ *   be sent if the file transfers completely.
+ * * filecancelled: Sent if the file was cancelled before it could finish.
+ * * fileerror: Sent if one of the file's parts failed to transfer. The event data
+ *   will include these additional elements:
+ *   * {Array} errors: List of errors that occurred to prevent the upload.
  */
-export default class DirectBinaryUploadProcess extends UploadOptionsBase {
-    constructor(options, uploadOptions) {
-        super(options, uploadOptions);
-
-        this.cancelled = false;
-        this.cancelledFiles = {};
-        this.cancelTokens = {};
-        this.initCancelToken = null;
+export default class DirectBinaryUploadProcess extends FileTransferHandler {
+    /**
+     * Constructs a new instance of the upload process for a single directory.
+     * @param {object} options Overall direct binary process options.
+     * @param {DirectBinaryUploadOptions} uploadOptions Options specific to the
+     *  current upload.
+     * @param {HttpClient} [httpClient] Client to use to submit HTTP requests. If
+     *  not specified then the process will create its own.
+     * @param {PartUpload} [partUploader] Upload to use when uploading file parts.
+     *  If not specified then the process will create its own.
+     */
+    constructor(options, uploadOptions, httpClient = false, partUploader = false) {
+        super(options, uploadOptions, httpClient);
+
+        this.partUploader = partUploader;
+        if (!this.partUploader) {
+            this.partUploader = new PartUploader(options, uploadOptions, this.getHttpClient(),
+                new ConcurrentQueue(options, uploadOptions));
+        }
+
         this.fileResults = {};
         this.fileEvents = {};
         this.fileTransfer = {};
+        this.completeUri = '';
     }
 
     /**
      * Does the work of uploading all files based on the upload options provided to the process.
      *
-     * @returns {Promise} Resolved with an UploadResult instance when the process has finished.
+     * @param {UploadResult} uploadResult Result to which information about the upload will be
+     *  added.
+     * @returns {Promise} Resolves when all files have been uploaded.
+     */
+    async upload(uploadResult) {
+        const initResponse = await this.initiateUpload(uploadResult);
+        this.completeUri = initResponse.getCompleteUri();
+
+        await this.partUploader.uploadParts(uploadResult, initResponse.getAllParts(), this);
+
+        uploadResult.stopTimer();
+
+        // output json result to logger
+        this.logInfo('Uploading result in JSON: ' + JSON.stringify(uploadResult, null, 4));
```

Comment:
If there are secrets, or presigned urls in the `uploadResult` object, you'll need to redact them.

Reply:
Good catch, thank you. I've removed any signed URLs from this operation.